# Лабораторная работа №5
##  Численное интегрирование таблично заданных функций.
---  
### Цель: Оценка точности численных методов интегрирования (метод трапеции, метод Симпсона, метод Рунге)
---
#### Для исследования выберем пункт VII.9.5(в): 
| x    | $ f(x) $|
| ---  | ---     |
| 0.00 | 1.00000 |
| 0.15 | 1.00757 |
| 0.30 | 1.03112 |
| 0.45 | 1.07346 |
| 0.60 | 1.14023 |
| 0.75 | 1.24213 |
| 0.90 | 1.40018 |
| 1.05 | 1.66030 |
| 1.20 | 2.14346 |

---
#### Метод трапеции: 
* $\displaystyle I_{h} =  h \cdot \frac{f_1 + 2 \cdot \sum\limits_{i=2}^{n - 1}{f_{i}} + f_{n}}{2}$

---
#### Метод Симпсона (при $ n = 9 $): 
* $\displaystyle I_{h} =  h \cdot \frac{f_1 + 4f_{2} + 2f_{3} + 4f_{4} + \dots + 2f_{7} + 4f_{8} + f_{9}}{3}$

---
#### Уточнение Рунге: 
* $\displaystyle I =  \frac{I_{h} - I_{rh}}{r^{p} - 1}$

---

#### Испорт модулей:

In [4]:
import numpy as np
from tabulate import tabulate

#### Объявление функций: 

In [5]:
def trapezoidal_rule(f, h, step=1):
    num = len(f)
    
    result = 0
    for i in range(0, num - step, step):
        result += (f[i] + f[i + step])
    
    return result * step * h / 2

def runge_rule(res_trapez_1, res_trapez_2):
    return (4 * res_trapez_1 - res_trapez_2)/(3)

def simpsons_rule(f, h):
    result = f[0]
    for i in range (1, len(f) - 2, 2):
        result += 4 * f[i] + 2 * f[i + 1]
    return (result + 4 * f[-2] + f[-1]) * h / 3

#### Основная часть

In [7]:
def main():

    print ("Input data:")
    table = [['x', 'f(x)'], 
        ['0.00', '1.000000'], 
        ['0.15', '1.007568'], 
        ['0.30', '1.031121'],
        ['0.45', '1.073456'],
        ['0.60', '1.140228'],
        ['0.75', '1.242129'],
        ['0.90', '1.400176'],
        ['1.05', '1.660300'],
        ['1.20', '2.143460'],
        ]

    print("----------------------------------")

    len_table = len(table)
    print(tabulate(table, headers='firstrow', tablefmt='fancy_grid'))

    x = []
    f = []

    for i in range (1, len_table):
        x.append(float(table[i][0]))
        f.append(float(table[i][1]))

    h = x[1] - x[0]

    print ("Let's calculate the integral with the trapezoidal rule...")
    res_trapez_1 = trapezoidal_rule(f, h)
    print("Result: " + str(res_trapez_1))
    print("----------------------------------")


    print("Let's calculate the integral with the trapezoidal rule with the double step... ")
    res_trapez_2 = trapezoidal_rule(f, h, 2)
    print("Result: " + str(res_trapez_2))
    print("----------------------------------")

    print("Let's clarify the previous integral with the runge rule... ")
    res_runge = runge_rule(res_trapez_1, res_trapez_2)
    print("Result: " + str(res_runge))
    print("----------------------------------")

    print("Let's calculate the integral with the simpson's rule... ")
    res_simpson = simpsons_rule(f, h)
    print("Result: " + str(res_simpson))
    print("----------------------------------")

    print("Conclusion:")
    print("Trapezoidal rule:", res_trapez_1)
    print("Trapezoidal rule with the double step:", res_trapez_2)
    print("Applying the runge rule:", res_runge)
    print("Simpson's rule:", res_simpson)

main()

Input data:
----------------------------------
╒══════╤═════════╕
│    x │    f(x) │
╞══════╪═════════╡
│ 0    │ 1       │
├──────┼─────────┤
│ 0.15 │ 1.00757 │
├──────┼─────────┤
│ 0.3  │ 1.03112 │
├──────┼─────────┤
│ 0.45 │ 1.07346 │
├──────┼─────────┤
│ 0.6  │ 1.14023 │
├──────┼─────────┤
│ 0.75 │ 1.24213 │
├──────┼─────────┤
│ 0.9  │ 1.40018 │
├──────┼─────────┤
│ 1.05 │ 1.6603  │
├──────┼─────────┤
│ 1.2  │ 2.14346 │
╘══════╧═════════╛
Let's calculate the integral with the trapezoidal rule...
Result: 1.5190062
----------------------------------
Let's calculate the integral with the trapezoidal rule with the double step... 
Result: 1.5429765
----------------------------------
Let's clarify the previous integral with the runge rule... 
Result: 1.5110161
----------------------------------
Let's calculate the integral with the simpson's rule... 
Result: 1.5110161
----------------------------------
Conclusion:
Trapezoidal rule: 1.5190062
Trapezoidal rule with the double step: 1.542976

## Заключение
### Для исследуемой таблично заданной функции посчитаны значения определенного интеграла всеми интересующими нас методами:

---

#### Результаты следующие:
* Trapezoidal rule: 1.5190062
* Trapezoidal rule with the double step: 1.5429765
* Applying the runge rule: 1.5110161
* Simpson's rule: 1.5110161
#### Отклоенения методов:

---

#### Трапеция и Симпсон: 
* $ \delta = 0.0079901 $
* $ \varepsilon \approx 0.5~\%$ 

---

#### Трапеция с удвоенной сеткой и Симпсон: 
* $ \delta = 0.0319604 $
* $ \varepsilon \approx 2~\%$ 

---

#### Трапеция и Рунге: 
* $ \delta = 0.0079901 $
* $ \varepsilon \approx 0.5~\%$ 

---

#### Рунге и Симпсон: 
* $ \delta = 0 $
* $ \varepsilon \approx 0~\%$ 

####
### По результатам проведенной работе было экспериментально показано, что значения интеграла, полученного методом Симпсона и уточнение Рунге совпадают. Отклонение же метода трапеции с одинарной и удвоенной сеткой от метода Симпсона невелико (порядка нескольких процентов)